common/src/main/java/com/discordsrv/common/feature/bansync/BanSyncModule.java

```diff
@@ -263,16 +276,39 @@ protected CompletableFuture<ISyncResult> applyDiscord(BanSyncConfig config, long
         }
 
         UserSnowflake snowflake = UserSnowflake.fromId(userId);
-        if (newState != null) {
-            return guild.ban(snowflake, config.discordMessageHoursToDelete, TimeUnit.HOURS)
-                    .reason(discordSRV.placeholderService().replacePlaceholders(config.discordBanReasonFormat, newState))
-                    .submit()
-                    .thenApply(v -> GenericSyncResults.ADD_DISCORD);
-        } else {
-            return guild.unban(snowflake)
-                    .reason(discordSRV.placeholderService().replacePlaceholders(config.discordUnbanReasonFormat))
-                    .submit()
-                    .thenApply(v -> GenericSyncResults.REMOVE_DISCORD);
+        switch (config.minecraftToDiscord.action) {
+            case BAN:
+                if (newState != null) {
+                    return guild.ban(snowflake, config.minecraftToDiscord.ban.messageHoursToDelete, TimeUnit.HOURS)
+                            .reason(discordSRV.placeholderService().replacePlaceholders(config.minecraftToDiscord.ban.banReasonFormat, newState))
+                            .submit()
+                            .thenApply(v -> GenericSyncResults.ADD_DISCORD);
+                } else {
+                    return guild.unban(snowflake)
+                            .reason(discordSRV.placeholderService().replacePlaceholders(config.minecraftToDiscord.ban.unbanReasonFormat))
+                            .submit()
+                            .thenApply(v -> GenericSyncResults.REMOVE_DISCORD);
+                }
+            case ROLE:
+                boolean isBan = newState != null;
+                return guild.retrieveMember(snowflake)
+                        .submit()
+                        .handle((member, t) -> {
+                            if (t instanceof ErrorResponseException && ((ErrorResponseException) t).getErrorResponse() == ErrorResponse.UNKNOWN_MEMBER)
+                                throw new SyncFail(BanSyncResult.NOT_A_GUILD_MEMBER);
+
+                            Set<Role> roles = new HashSet<>();
+
+                            roles.add(guild.getRoleById(config.minecraftToDiscord.role.roleId));
+
+                            return guild.modifyMemberRoles(member, isBan ? roles : Collections.emptySet(), isBan ? Collections.emptySet() : roles)
+                                    .reason("DiscordSRV ban synchronization")
+                                    .submit();
+                        })
+                        .thenCompose(r -> r) // Flatten the completablefuture
+                        .thenApply(v -> isBan ? GenericSyncResults.ADD_DISCORD : GenericSyncResults.REMOVE_DISCORD);
+            default:
+                return CompletableFutureUtil.failed(new SyncFail(BanSyncResult.NO_DISCORD_CONNECTION));
```

Comment:
Result should be invalid config

Reply:
No memory of writing that. Maybe I copied it from somewhere 🤦‍♂️